src/fu-agent.c

```diff
@@ -0,0 +1,645 @@
+/*
+ * Copyright (C) 2019 Richard Hughes <richard@hughsie.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#define G_LOG_DOMAIN				"FuAgent"
+
+#include "config.h"
+
+#include <fwupd.h>
+#include <glib/gi18n.h>
+#include <glib-unix.h>
+#include <locale.h>
+#include <stdlib.h>
+#include <unistd.h>
+
+#include "fu-common.h"
+#include "fu-util-common.h"
+#include "fwupd-device-private.h"
+
+struct FuUtilPrivate {
+	GCancellable		*cancellable;
+	GMainLoop		*loop;
+	GOptionContext		*context;
+	SoupSession		*soup_session;
+	FwupdClient		*client;
+};
+
+static gchar *
+fu_util_agent_get_config_fn (void)
+{
+	g_autofree gchar *path = fu_common_get_path (FU_PATH_KIND_SYSCONFDIR_PKG);
+	return g_build_filename (path, "agent.conf", NULL);
+}
+
+static GKeyFile *
+fu_util_agent_get_config (GError **error)
+{
+	g_autofree gchar *fn = fu_util_agent_get_config_fn ();
+	g_autoptr(GKeyFile) config = g_key_file_new ();
+	if (!g_key_file_load_from_file (config, fn, G_KEY_FILE_KEEP_COMMENTS, error)) {
+		g_prefix_error (error, "failed to load %s: ", fn);
+		return FALSE;
+	}
+	return g_steal_pointer (&config);
+}
+
+static gboolean
+fu_util_agent_set_server (FuUtilPrivate *priv, const gchar *server, GError **error)
+{
+	g_autofree gchar *fn = fu_util_agent_get_config_fn ();
+	g_autoptr(GKeyFile) config = fu_util_agent_get_config (error);
+	if (config == NULL)
+		return FALSE;
+	g_key_file_set_string (config, "fwupdagent", "Server", server);
+	return g_key_file_save_to_file (config, fn, error);
+}
+
+static gboolean
+fu_util_agent_run_action (FuUtilPrivate *priv,
+			  JsonObject *json_object,
+			  GError **error)
+{
+	const gchar *checksum;
+	const gchar *device_id;
+	const gchar *task;
+	g_autoptr(FwupdDevice) device = NULL;
+	g_autoptr(FwupdRelease) rel = NULL;
+	g_autoptr(GPtrArray) releases = NULL;
+
+	/* check object for args */
+	if (!json_object_has_member (json_object, "Task")) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_INTERNAL,
+				     "No task specified");
+		return FALSE;
+	}
+	if (!json_object_has_member (json_object, "DeviceId")) {
```

Comment:
`FWUPD_RESULT_KEY_DEVICE_ID` maybe?

Reply:
Agreed, and `FWUPD_RESULT_KEY_CHECKSUM`.